salt/modules/boto_route53.py

```diff
@@ -651,44 +651,71 @@ def delete_record(name, zone, record_type, identifier=None, all_records=False,
+        current = change.GetChangeResponse.ChangeInfo.Status
+        if current == 'INSYNC':
             return True
-        i = i + 1
-        time.sleep(20)
-    log.error('Timed out waiting for Route53 status update.')
+        sleep = wait if wait % 60 == wait else 60
+        log.info('Sleeping {0} seconds waiting for changes to synch (current status {1})'.format(
+                 sleep, current))
+        time.sleep(sleep)
+        wait -= sleep
+        continue
+    log.error('Route53 changes not synced after {0} seconds.'.format(orig_wait))
     return False
 
 
-def create_hosted_zone(domain_name, caller_ref=None, comment='',
-                       private_zone=False, vpc_id=None, vpc_name=None,
-                       vpc_region=None, region=None, key=None, keyid=None,
+def create_hosted_zone(domain_name, caller_ref=None, comment='', private_zone=False, vpc_id=None,
+                       vpc_name=None, vpc_region=None, region=None, key=None, keyid=None,
                        profile=None):
     '''
-    Create a new Route53 Hosted Zone. Returns a Python data structure with
-    information about the newly created Hosted Zone.
+    Create a new Route53 Hosted Zone. Returns a Python data structure with information about the
+    newly created Hosted Zone.
 
     domain_name
-        The name of the domain. This should be a fully-specified domain, and
-        should terminate with a period. This is the name you have registered
-        with your DNS registrar. It is also the name you will delegate from your
-        registrar to the Amazon Route 53 delegation servers returned in response
+        The name of the domain. This must be fully-qualified, terminating with a period.  This is
+        the name you have registered with your domain registrar.  It is also the name you will
+        delegate from your registrar to the Amazon Route 53 delegation servers returned in response
         to this request.
 
     caller_ref
-        A unique string that identifies the request and that allows
-        create_hosted_zone() calls to be retried without the risk of executing
-        the operation twice.  You want to provide this where possible, since
-        additional calls while the first is in PENDING status will be accepted
-        and can lead to multiple copies of the zone being created in Route53.
+        A unique string that identifies the request and that allows create_hosted_zone() calls to
+        be retried without the risk of executing the operation twice.  It can take several minutes
+        for the change to replicate globally, and change from PENDING to INSYNC status. Thus it's
+        best to provide some value for this where possible, since duplicate calls while the first
+        is in PENDING status will be accepted and can lead to multiple copies of the zone being
+        created.  On the other hand, if a zone is created with a given caller_ref, then deleted,
```

Comment:
Ugh. So, the situation is the same as it was previously, where if someone doesn't use caller_ref (which is likely), then we may create multiple hosted zones. This isn't good default behavior. Why can't we generate some stable identifier automatically based on info being sent into the module, rather than generating a uuid?

Reply:
This is in the module function.  The STATE function manages things with a UUID, and a "wait for PENDING" loop, which between them solve both the original problem, and the one raised in #40655.  Exec module functions are historically much more "hands on" than state functions.  If someone runs this by hand, they'll get a meaningful error back, which combined with the verbiage I added should be enough to get them going.

We can't generate a stable identifier because that's what it ORIGINALLY did, which led to #40655 :)  If you pass in a "stable" caller_ref, you're back to the issue of "create / delete / create" cycles failing, while if you generate a non-stable caller_ref, you run the risk of duplicate zones, IFF you run create_hosted_zone() again against the same zone, and before the first call has reached INSYNC status.  Thus the addition of a "wait for INSYNC" loop to the state module.

Reply:
Ah. Gotcha. So in normal operation, it's going to be a change that waits until it's done. If someone runs the state again while this is happening you'll get a duplicate, but won't otherwise? That works for me.